### PR TITLE
feat: timeslip prototype added

### PR DESCRIPTION
**Pull request by monfera**
## Summary

Adds the timeslip prototype as a pre-alpha chart type. It's still mostly JS and prototypey, with partly functional dat.gui controls, lacking a tooltip and event callback integration. However it already uses the time raster and chrono utility from Cartesians (those have already been integrated by `xy_chart` from the original timeslip [prototype](https://timeslip.vercel.app/)).

image

The intent is to eventually unify what it provides with the production Cartesian charts.

An also discussed, shorter term parallel track is to refactor, typescriptify and round out the implementation and design so it serves as a specifically designed time series chart with cohesive design elements, which shares some attributes with bar charts incl. histograms, line charts and area charts.

## Details

There's a recognized tension between generative chart making, such as devising a form of grammar of graphics (significantly generative approaches include Vega / Vega-Lite, Charticulator, Sanddance, some aspects of Tableau) and fixed chart types (Excel, most aspects of Plotly and `elastic-charts`, some aspects of Tableau). This prototype is more of a recognition that it's a challenging design and implementation task to even approach a small point in the design space, and is a source of further work (eg. in the generative direction), rather than vouching for the supremacy of a rigid, chart type based approach.

However it is often the case, eg. in data discovery mode, that zoom and pan, interactive re-binning etc. are already valuable, while there's no current compositioning need. It is also often good practice to just visualize a single time series in a chart, and to not imply a possibly untrue degree of continuity of the time series values (which regular line and area charts do).

List of features, some of which are slated for Cartesian integration:

1. Multi-layer time axis and time grid (already [merged into](https://github.com/elastic/elastic-charts/pull/1326) `elastic-charts` Cartesians
2. Redundant encoding of time units: besides the calendar-style time axis, the bin width (line section with unit labeling) and domain extent are shown atop of the chart
3. Zoom dependent data re-discretization into bins (compliantly binned data comes from outside Charts, but driven on demand by the zoom level and pan position); unreasonably wide bars and "needle bars" are both avoided
4. Semantic styling options: instead of asking users to pass weights/colors etc, they just specify high vs normal quantitative readability etc.
5. High FPS zoom 
6. Kinetic panning
7. Smooth Y domain fitting on zoom, pan and re-binning (and in the future, upon in-place data changes and incremental fetch)
8. Fading partial bars in/out at the edge of the chart 
9. Staggering and fading time tick labels in-out at the edges 
10. Reinforcing line at the distal end of the bar for increased contrast and accessibility, without needing to shade large areas with an overly high contrast (the reinforcing line need to become a bit more salient, and the zero line might switch from dark gray to blue as well)
11. Manually balanced saliency hierarchy of the grid line weights (for dark and light backgrounds)
12. Design for avoiding artifacts among bars, and between bars and grid lines (for example, no bar gaps)
13. Internationalization (see the corresponding dat.gui options on the top right)
14. Redundant Y axes, esp. useful in wide aspect ratio time series charts (Cartesian charts can already do this)

### Checklist

- [x] The proper **chart type** label has been added (e.g. `:xy`, `:partition`)
- [x] The proper **feature** labels have been added (e.g. `:interactions`, `:axis`)
- [ ] The `:theme` label has been added and the `@elastic/eui-design` team has been pinged when there are `Theme` API changes
- [ ] All related issues have been linked (i.e. `closes #123`, `fixes #123`)
- [ ] New public API exports have been added to `packages/charts/src/index.ts`
- [ ] Unit tests have been added or updated to match the most common scenarios
- [ ] The proper documentation and/or storybook story has been added or updated
- [ ] The code has been checked for cross-browser compatibility (Chrome, Firefox, Safari, Edge)
- [ ] Visual changes have been tested with all available themes including `dark`, `light`, `eui-dark` & `eui-light`

### CODE REVIEW

**Comment by nickofthyme** [comment]
buildkite update vrt

**Comment by nickofthyme** [comment]
buildkite update vrt

File: storybook/stories/timeslip/01_timeslip.story.tsx [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  const theme: PartialTheme = {
+    chartMargins: { top: 0, left: 0, bottom: 0, right: 0 },
+    chartPaddings: { left: 0, right: 0, top: 0, bottom: 0 },
+  };
+  button('Reset focus', () => {
+    resetFocusControl();
+  });
+  button('Set focus on random node', () => {
+    focusOnNodeControl(Math.floor(20 * Math.random()));
+  });
```
Comment (markov00) [inline storybook/stories/timeslip/01_timeslip.story.tsx:68]: You can remove this

File: storybook/stories/timeslip/01_timeslip.story.tsx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+export const Example = () => {
+  const resetFocusControl: FlameGlobalControl = noop; // initial value
+  const focusOnNodeControl: FlameNodeControl = noop; // initial value
+
+  const onElementListeners = {
+    onElementClick: action('onElementClick'),
+    onElementOver: action('onElementOver'),
+    onElementOut: action('onElementOut'),
+  };
```
Comment (markov00) [inline storybook/stories/timeslip/01_timeslip.story.tsx:58]: Also those are not currently used

File: packages/charts/package.json [hunk 3]
```diff
@@ -36,6 +36,7 @@
     "bezier-easing": "^2.1.0",
     "chroma-js": "^2.1.0",
     "classnames": "^2.2.6",
+    "dat.gui": "^0.7.9",
```
Comment (markov00) [inline packages/charts/package.json:39]: This should be moved as a dependency in `storybook/package.json`

[the remark this answers is not in the document]
Comment (monfera) [reply packages/charts/package.json:39]: We landed on the decision to remove `dat.gui` and in-story configurability altogether. Proper, chart config based settings and story controls can be added in a future PR if needed

Comment (monfera) [reply packages/charts/package.json:39]: See above, dat.gui will be removed

[the remark this answers is not in the document]
Comment (monfera) [reply packages/charts/package.json]: b08ba2d91bdfce11c695487d73d318baa050131f

Comment (monfera) [reply packages/charts/package.json]: b08ba2d91bdfce11c695487d73d318baa050131f

File: storybook/stories/timeslip/01_timeslip.story.tsx (hunk 2 above)
Comment (monfera) [reply storybook/stories/timeslip/01_timeslip.story.tsx]: 4047b5e6410bceb8bac01b5c3f5482bc33d1861c

**Review by markov00 — APPROVED** [review]
LGTM

**Pull request merged**